mesh/mesh_operators.hpp

```diff
@@ -422,6 +422,166 @@ class CoefficientRefiner : public MeshOperator
    virtual void Reset();
 };
 
+/** @brief Mesh refinement using Maximum Marking
+
+  This class uses the given ErrorEstimator to estimate local element errors
+  and then marks for refinement all elements i such that loc_err_i > threshold.
+  The threshold is computed as
+  \code
+  threshold = gamma * Max_{i in Mesh} { local_error_i }
+  \endcode
+  where gamma is a configurable parameter
+  */
+class MaximumMarkingRefiner : public MeshOperator
+{
+protected:
+   ErrorEstimator &estimator;
+
+   long   max_elements;
+
+   double gamma;
+   double threshold;
+   long num_marked_elements;
+
+   Array<Refinement> marked_elements;
+   long current_sequence;
+
+   int non_conforming;
+   int nc_limit;
+
+   /** @brief Apply the operator to the mesh.
+     @return STOP if a stopping criterion is satisfied or no elements were
+     marked for refinement; REFINED + CONTINUE otherwise. */
+   virtual int ApplyImpl(Mesh &mesh) override;
+
+public:
+   /// Construct a MaximumMarkingRefiner using the given ErrorEstimator.
+   MaximumMarkingRefiner(ErrorEstimator &est);
+
+   // default destructor (virtual)
+
+   void SetMaxElements(long max_elem) { max_elements = max_elem; }
+
+   /// Use nonconforming refinement, if possible (triangles, quads, hexes).
+   void PreferNonconformingRefinement() { non_conforming = 1; }
+
+   /** @brief Use conforming refinement, if possible (triangles, tetrahedra)
+     -- this is the default. */
+   void PreferConformingRefinement() { non_conforming = -1; }
+
+   /** @brief Set the maximum ratio of refinement levels of adjacent elements
+     (0 = unlimited). */
+   void SetNCLimit(int nc_limit_)
+   {
+      MFEM_ASSERT(nc_limit_ >= 0, "Invalid NC limit");
+      nc_limit = nc_limit_;
+   }
+
+   /// Get the number of marked elements in the last Apply() call.
+   long GetNumMarkedElements() const { return num_marked_elements; }
+
+   /// Set the confiugrable parameter gamma
+   void SetGamma( double theta ) { gamma = theta; }
+   /// Get the configured gamma
+   double GetGamma() const { return gamma; }
+
+   /// Get the threshold used in the last Apply() call.
+   double GetThreshold() const { return threshold; }
+
+   /// Reset the associated estimator.
+   virtual void Reset() override
+   {
+      estimator.Reset();
+      current_sequence = -1;
+      num_marked_elements = 0;
+   };
+};
+
+/** @brief Mesh refinement using Dörfler Marking
+
+  This class uses the given ErrorEstimator to estimate local element errors
+  and then marks for refinement a set M such that
```

Comment:
Sorry for the question, but isn't the point of Dörfler Marking to find a set M with minimal cardinality?